File: 2019-09-28/CygniParser/src/cygni/parser/Parser.java
package cygni.parser;

import cygni.ast.Module;
import cygni.lexer.Tag;
import cygni.lexer.Token;
import cygni.exceptions.ParserException;
import cygni.ast.*;
import cygni.types.*;

import java.util.ArrayList;

public class Parser {
    private ArrayList<Token> tokens;
    private int offset;
    private String path;

    public Parser(String path, ArrayList<Token> tokens) {
        this.tokens = tokens;
        this.offset = 0;
        this.path = path;
    }

    public Module module() throws ParserException {
        int startLine = look().line;
        int startCol = look().col;
        String moduleName;
        ArrayList<String> names = new ArrayList<>();
        if (look().tag == Tag.Module) {
            match(Tag.Module);
            moduleName = (String) match(Tag.Identifier).value;
            match(Tag.LeftParenthesis);
            while (!eof() && look().tag != Tag.RightParenthesis) {
                names.add((String) match(Tag.Identifier).value);
            }
            match(Tag.RightParenthesis);
        } else {
            moduleName = "(default)";
        }
        ArrayList<Node> nodes = new ArrayList<Node>();
        while (!eof()) {
            nodes.add(statement());
        }
        return new Module(startLine, startCol, look().line, look().col, path, moduleName, names, nodes);
    }

    private boolean eof() {
        return look().tag == Tag.Eof;
    }

    private Token look() {
        return tokens.get(offset);
    }

    private void move() {
        this.offset++;
    }

    private Token match(Tag tag) throws ParserException {
        if (tag == look().tag) {
            Token t = look();
            move();
            return t;
        } else {
            throw new ParserException(look().line, look().col,
                    "expecting " + tag + ", got " + look().tag);
        }
    }

    private Node statement() throws ParserException {
        if (look().tag == Tag.If) {
            return ifStatement();
        } else if (look().tag == Tag.Var) {
            return var();
        } else if (look().tag == Tag.Def) {
            return def();
        } else if (look().tag == Tag.Return) {
            return parseReturn();
        } else if (look().tag == Tag.While) {
            return parseWhile();
        } else {
            return assign();
        }
    }

    private Node assign() throws ParserException {
        int startLine = look().line;
        int startCol = look().col;
        Node x = or();
        if (look().tag == Tag.Assign) {
            Token t = match(Tag.Assign);
            Node y = or();
            return new Assign(startLine, startCol, look().line, look().col, x, y);
        } else {
            return x;
        }
    }

    private Node or() throws ParserException {
        int startLine = look().line;
        int startCol = look().col;
        Node x = and();
        while (look().tag == Tag.Or) {
            Token t = match(Tag.Or);
            Node y = and();
            x = new BinaryOp(startLine, startCol, look().line, look().col, BinaryOpKind.Or, x, y);
        }
        return x;
    }

    private Node and() throws ParserException {
        int startLine = look().line;
        int startCol = look().col;
        Node x = equality();
        while (look().tag == Tag.And) {
            Token t = match(Tag.And);
            Node y = equality();
            x = new BinaryOp(startLine, startCol, look().line, look().col, BinaryOpKind.And, x, y);
        }
        return x;
    }

    private Node equality() throws ParserException {
        int startLine = look().line;
        int startCol = look().col;
        Node x = relation();
        while (look().tag == Tag.Equal || look().tag == Tag.NotEqual) {
            Token t = look();
            move();
            Node y = relation();
            if (t.tag == Tag.Equal) {
                x = new BinaryOp(startLine, startCol, look().line, look().col, BinaryOpKind.Equal, x, y);
            } else {
                x = new BinaryOp(startLine, startCol, look().line, look().col, BinaryOpKind.NotEqual, x, y);
            }
        }
        return x;
    }

    private Node relation() throws ParserException {
        int startLine = look().line;
        int startCol = look().col;
        Node x = expr();
        if (look().tag == Tag.GreaterThan
                || look().tag == Tag.LessThan
                || look().tag == Tag.GreaterThanOrEqual
                || look().tag == Tag.LessThanOrEqual) {
            Token t = look();
            move();
            Node y = expr();
            if (t.tag == Tag.GreaterThan) {
                return new BinaryOp(startLine, startCol, look().line, look().col, BinaryOpKind.GreaterThan, x, y);
            } else if (t.tag == Tag.LessThan) {
                return new BinaryOp(startLine, startCol, look().line, look().col, BinaryOpKind.LessThan, x, y);
            } else if (t.tag == Tag.GreaterThanOrEqual) {
                return new BinaryOp(startLine, startCol, look().line, look().col, BinaryOpKind.GreaterThanOrEqual, x, y);
            } else {
                return new BinaryOp(startLine, startCol, look().line, look().col, BinaryOpKind.LessThanOrEqual, x, y);
            }
        } else {
            return x;
        }
    }

    private Node expr() throws ParserException {
        int startLine = look().line;
        int startCol = look().col;
        Node x = term();
        while (look().tag == Tag.Add || look().tag == Tag.Subtract) {
            Token t = look();
            move();
            Node y = term();
            if (t.tag == Tag.Add) {
                x = new BinaryOp(startLine, startCol, look().line, look().col, BinaryOpKind.Add, x, y);
            } else {
                x = new BinaryOp(startLine, startCol, look().line, look().col, BinaryOpKind.Subtract, x, y);
            }
        }
        return x;
    }

    private Node term() throws ParserException {
        int startLine = look().line;
        int startCol = look().col;
        Node x = unary();
        while (look().tag == Tag.Multiply || look().tag == Tag.Divide) {
            Token t = look();
            move();
            Node y = unary();
            if (t.tag == Tag.Multiply) {
                x = new BinaryOp(startLine, startCol, look().line, look().col, BinaryOpKind.Multiply, x, y);
            } else {
                x = new BinaryOp(startLine, startCol, look().line, look().col, BinaryOpKind.Divide, x, y);
            }
        }
        return x;
    }

    private Node unary() throws ParserException {
        int startLine = look().line;
        int startCol = look().col;
        if (look().tag == Tag.Subtract) {
            move();
            Node x = unary();
            return new UnaryOp(startLine, startCol, look().line, look().col, UnaryOpKind.Negate, x);
        } else if (look().tag == Tag.Not) {
            move();
            Node x = unary();
            return new UnaryOp(startLine, startCol, look().line, look().col, UnaryOpKind.Not, x);
        } else {
            return postfix();
        }
    }

    private Node postfix() throws ParserException {
        Node x = factor();
        while (look().tag == Tag.LeftParenthesis || look().tag == Tag.LeftBracket) {
            int startLine = look().line;
            int startCol = look().col;
            if (look().tag == Tag.LeftParenthesis) {
                ArrayList<Node> arguments = new ArrayList<Node>();
                match(Tag.LeftParenthesis);
                if (look().tag == Tag.RightParenthesis) {
                    match(Tag.RightParenthesis);
                    x = new Call(startLine, startCol, look().line, look().col, x, arguments);
                } else {
                    arguments.add(or());
                    while (!eof() && look().tag != Tag.RightParenthesis) {
                        match(Tag.Comma);
                        arguments.add(or());
                    }
                    match(Tag.RightParenthesis);
                    x = new Call(startLine, startCol, look().line, look().col, x, arguments);
                }
            } else if (look().tag == Tag.LeftBracket) {
                ArrayList<Type> types = parseTypeArguments();
                x = new Specialize(startLine, startCol, look().line, look().col, x, types);
            }
        }
        return x;
    }

    private Node factor() throws ParserException {
        if (look().tag == Tag.LeftParenthesis) {
            move();
            Node x = or();
            match(Tag.RightParenthesis);
            return x;
        } else if (look().tag == Tag.Integer || look().tag == Tag.Float || look().tag == Tag.String) {
            Object v = look().value;
            int startLine = look().line;
            int startCol = look().col;
            move();
            Node x = new Constant(startLine, startCol, look().line, look().col, v);
            return x;
        } else if (look().tag == Tag.True) {
            int startLine = look().line;
            int startCol = look().col;
            move();
            Node x = new Constant(startLine, startCol, look().line, look().col, true);
            return x;
        } else if (look().tag == Tag.False) {
            int startLine = look().line;
            int startCol = look().col;
            move();
            Node x = new Constant(startLine, startCol, look().line, look().col, false);
            return x;
        } else if (look().tag == Tag.Identifier) {
            String name = (String) look().value;
            int startLine = look().line;
            int startCol = look().col;
            move();
            Node x = new Name(startLine, startCol, look().line, look().col, name);
            return x;
        } else if (look().tag == Tag.LeftBracket) {
            return initArray();
        } else {
            throw new ParserException(look().line, look().col, "factor " + look().tag);
        }
    }

    private Node block() throws ParserException {
        int startLine = look().line;
        int startCol = look().col;
        match(Tag.LeftBrace);
        ArrayList<Node> nodes = new ArrayList<Node>();
        while (!eof() && look().tag != Tag.RightBrace) {
            nodes.add(statement());
        }
        match(Tag.RightBrace);
        return new Block(startLine, startCol, look().line, look().col, nodes);
    }

    private Node ifStatement() throws ParserException {
        int startLine = look().line;
        int startCol = look().col;
        match(Tag.If);
        Node condition = or();
        Node ifTrue = block();
        if (look().tag == Tag.Else) {
            match(Tag.Else);
            if (look().tag == Tag.If) {
                Node chunk = ifStatement();
                return new IfElse(startLine, startCol, look().line, look().col, condition, ifTrue, chunk);
            } else {
                Node chunk = block();
                return new IfElse(startLine, startCol, look().line, look().col, condition, ifTrue, chunk);
            }
        } else {
            return new IfThen(startLine, startCol, look().line, look().col, condition, ifTrue);
        }
    }

    private Node var() throws ParserException {
        int startLine = look().line;
        int startCol = look().col;
        match(Tag.Var);
        Token t = match(Tag.Identifier);
        match(Tag.Assign);
        Node value = or();
        return new Var(startLine, startCol, look().line, look().col, (String) t.value, value);
    }

    private Node def() throws ParserException {
        int startLine = look().line;
        int startCol = look().col;
        match(Tag.Def);
        String name = (String) match(Tag.Identifier).value;
        ArrayList<UnknownType> unknownTypes = new ArrayList<UnknownType>();
        if (look().tag == Tag.LeftBracket) {
            unknownTypes = parseGenericTypesDef();
        }
        match(Tag.LeftParenthesis);
        ArrayList<Parameter> parameters = new ArrayList<Parameter>();
        if (look().tag != Tag.RightParenthesis) {
            parameters.add(parameter());
            while (!eof() && look().tag != Tag.RightParenthesis) {
                match(Tag.Comma);
                parameters.add(parameter());
            }
        }
        match(Tag.RightParenthesis);
        match(Tag.Colon);
        Type returnType = parseType();
        Node body = block();
        return new Def(startLine, startCol, look().line, look().col, name, parameters, returnType, body, unknownTypes);
    }

    private Parameter parameter() throws ParserException {
        String name = (String) match(Tag.Identifier).value;
        match(Tag.Colon);
        Type type = parseType();
        return new Parameter(name, type);
    }


    private Type parseType() throws ParserException {
        String name = (String) match(Tag.Identifier).value;
        if (look().tag == Tag.LeftBracket) {
            ArrayList<Type> types = parseTypeArguments();
            Type result = Type.makeType(name, types);
            if (result == null) {
                throw new ParserException(look().line, look().col, "type error");
            } else {
                return result;
            }
        } else {
            return Type.makePrimitiveType(name);
        }
    }

    private Node parseReturn() throws ParserException {
        int startLine = look().line;
        int startCol = look().col;
        match(Tag.Return);
        Node value = or();
        return new Return(startLine, startCol, look().line, look().col, value);
    }

    private ArrayList<Type> parseTypeArguments() throws ParserException {
        match(Tag.LeftBracket);
        ArrayList<Type> types = new ArrayList<Type>();
        types.add(parseType());
        while (!eof() && look().tag != Tag.RightBracket) {
            match(Tag.Comma);
            types.add(parseType());
        }
        match(Tag.RightBracket);
        return types;
    }

    private Node initArray() throws ParserException {
        int startLine = look().line;
        int startCol = look().col;
        match(Tag.LeftBracket);
        ArrayList<Node> elements = new ArrayList<Node>();
        if (look().tag == Tag.RightBracket) {
            return new InitArray(startLine, startCol, look().line, look().col, elements);
        } else {
            elements.add(or());
            while (!eof() && look().tag != Tag.RightBracket) {
                match(Tag.Comma);
                elements.add(or());
            }
            match(Tag.RightBracket);
            return new InitArray(startLine, startCol, look().line, look().col, elements);
        }
    }

    private ArrayList<UnknownType> parseGenericTypesDef() throws ParserException {
        match(Tag.LeftBracket);
        ArrayList<UnknownType> unknownTypes = new ArrayList<UnknownType>();
        unknownTypes.add(parseOneGenericTypeDef());
        while (!eof() && look().tag != Tag.RightBracket) {
            match(Tag.Comma);
            unknownTypes.add(parseOneGenericTypeDef());
        }
        match(Tag.RightBracket);
        return unknownTypes;
    }

    private UnknownType parseOneGenericTypeDef() throws ParserException {
        String name = (String) match(Tag.Identifier).value;
        return new UnknownType(name);
    }

    private While parseWhile() throws ParserException {
        Token token = match(Tag.While);
        Node condition = or();
        Node body = block();
        return new While(token.line, token.col, look().line, look().col, condition, body);
    }

    private DefClass defClass() throws ParserException {
        match(Tag.Class);
        String name = (String) match(Tag.Identifier).value;
        ArrayList<UnknownType> unknownTypes;
        if (look().tag == Tag.LeftBracket) {
            unknownTypes = parseGenericTypesDef();
        } else {
            unknownTypes = new ArrayList<>();
        }
        ArrayList<DefClass.Field> fields = new ArrayList<>();
        ArrayList<DefClass.Method> methods = new ArrayList<>();
        match(Tag.LeftBrace);
        while (!eof() && look().tag != Tag.RightBrace) {
            DefClass.Access access;
            if (look().tag == Tag.Private) {
                access = DefClass.Access.Private;
            } else if (look().tag == Tag.Protected) {
                access = DefClass.Access.Protected;
            } else {
                access = DefClass.Access.Public;
            }

            if (look().tag == Tag.Var) {
                match(Tag.Var);
                String varName = (String) match(Tag.Identifier).value;
                match(Tag.Colon);
                Type
            }
        }
    }
}

File: Cygni/src/sources/Parser.cpp
#include "Parser.hpp"
using std::vector;
using std::unordered_map;
namespace cygni {
Parser::Parser(std::vector<Token> tokens, std::shared_ptr<SourceDocument> document)
    :tokens{tokens}, document{document}{
    }


Program Parser::ParseProgram() {
    std::unordered_map<std::u32string, std::shared_ptr<ClassInfo>> classes;
  while (!IsEof()) {
    if (Look().tag == Tag::Class) {
        ParseDefClass();
    } else {
        ParseDefModule();
    }
  }
  std::u32string path;
  return Program(path, classes);
}

ExpPtr Parser::Statement() {
  switch (Look().tag) {
    case Tag::If:
      return IfStatement();
    case Tag::Var:
      return ParseVar();
    case Tag::Def:
      return ParseDef();
    case Tag::Return:
      return ParseReturn();
    case Tag::While:
      return ParseWhile();
    default:
      return ParseAssign();
  }
}

ExpPtr Parser::ParseAssign() {
  const Token& start = Look();
  auto x = ParseOr();
  if (Look().tag == Tag::Assign) {
    Token t = Match(Tag::Assign);
    auto y = ParseOr();
    return std::make_shared<BinaryExpression>(GetLoc(start), ExpressionType::Assign, x, y);
  } else {
    return x;
  }
}

ExpPtr Parser::ParseOr() {
  const Token& start = Look();
  auto x = ParseAnd();
  while (Look().tag == Tag::Or) {
    Token t = Match(Tag::Or);
    auto y = ParseAnd();
    x = std::make_shared<BinaryExpression>(GetLoc(start), ExpressionType::Or, x, y);
  }
  return x;
}

ExpPtr Parser::ParseAnd() {
  const Token& start = Look();
  auto x = ParseEquality();
  while (Look().tag == Tag::And) {
    Token t = Match(Tag::And);
    auto y = ParseEquality();
    x = std::make_shared<BinaryExpression>(GetLoc(start), ExpressionType::And, x, y);
  }
  return x;
}

ExpPtr Parser::ParseEquality() {
  const Token& start = Look();
  auto x = ParseRelation();
  while (Look().tag == Tag::Equal || Look().tag == Tag::NotEqual) {
    Token t = Look();
    Advance();
    auto y = ParseRelation();
    if (t.tag == Tag::Equal) {
      x = std::make_shared<BinaryExpression>(GetLoc(start), ExpressionType::Equal, x, y);
    } else {
      x = std::make_shared<BinaryExpression>(GetLoc(start), ExpressionType::NotEqual, x, y);
    }
  }
  return x;
}

ExpPtr Parser::ParseRelation() {
  const Token& start = Look();
  auto x = ParseExpr();
  if (Look().tag == Tag::GreaterThan || Look().tag == Tag::LessThan ||
      Look().tag == Tag::GreaterThanOrEqual ||
      Look().tag == Tag::LessThanOrEqual) {
    Token t = Look();
    Advance();
    auto y = ParseExpr();
    if (t.tag == Tag::GreaterThan) {
      return std::make_shared<BinaryExpression>(GetLoc(start), ExpressionType::GreaterThan, x, y);
    } else if (t.tag == Tag::LessThan) {
      return std::make_shared<BinaryExpression>(GetLoc(start), ExpressionType::LessThan, x, y);
    } else if (t.tag == Tag::GreaterThanOrEqual) {
      return std::make_shared<BinaryExpression>(GetLoc(start), ExpressionType::GreaterThanOrEqual, x, y);
    } else {
      return std::make_shared<BinaryExpression>(GetLoc(start), ExpressionType::LessThanOrEqual, x, y);
    }
  } else {
    return x;
  }
}

ExpPtr Parser::ParseExpr() {
  const Token& start = Look();
  auto x = ParseTerm();
  while (Look().tag == Tag::Add || Look().tag == Tag::Subtract) {
    Token t = Look();
    Advance();
    auto y = ParseTerm();
    if (t.tag == Tag::Add) {
      x = std::make_shared<BinaryExpression>(GetLoc(start), ExpressionType::Add, x, y);
    } else {
      x = std::make_shared<BinaryExpression>(GetLoc(start), ExpressionType::Subtract, x, y);
    }
  }
  return x;
}

ExpPtr Parser::ParseTerm() {
  const Token& start = Look();
  auto x = ParseUnary();
  while (Look().tag == Tag::Multiply || Look().tag == Tag::Divide) {
    Token t = Look();
    Advance();
    auto y = ParseUnary();
    if (t.tag == Tag::Multiply) {
      x = std::make_shared<BinaryExpression>(GetLoc(start), ExpressionType::Multiply, x, y);
    } else {
      x = std::make_shared<BinaryExpression>(GetLoc(start), ExpressionType::Divide, x, y);
    }
  }
  return x;
}

ExpPtr Parser::ParseUnary() {
  const Token& start = Look();
  if (Look().tag == Tag::Add) {
    Advance();
    auto x = ParseUnary();
    return std::make_shared<UnaryExpression>(GetLoc(start), ExpressionType::UnaryPlus, x);
  } else if (Look().tag == Tag::Subtract) {
    Advance();
    auto x = ParseUnary();
    return std::make_shared<UnaryExpression>(GetLoc(start), ExpressionType::UnaryMinus, x);
  } else if (Look().tag == Tag::Not) {
    Advance();
    auto x = ParseUnary();
    return std::make_shared<UnaryExpression>(GetLoc(start), ExpressionType::Not, x);
  } else {
    return ParsePostfix();
  }
}

ExpPtr Parser::ParsePostfix() {
  auto x = ParseFactor();
  while (Look().tag == Tag::LeftParenthesis || Look().tag == Tag::LeftBracket) {
    const Token& start = Look();

    if (Look().tag == Tag::LeftParenthesis) {
    vector<ExpPtr> arguments;
      Match(Tag::LeftParenthesis);
      if (Look().tag == Tag::RightParenthesis) {
        Match(Tag::RightParenthesis);
        x = std::make_shared<InvocationExpression>(GetLoc(start), x, arguments);
      } else {
        arguments.push_back(ParseOr());
        while (!IsEof() && Look().tag != Tag::RightParenthesis) {
          Match(Tag::Comma);
          arguments.push_back(ParseOr());
        }
        Match(Tag::RightParenthesis);
        x = std::make_shared<InvocationExpression>(GetLoc(start), x, arguments);
      }
    } else if (Look().tag == Tag::LeftBracket) {
      // TO DO
    }
  }
  return x;
}

ExpPtr Parser::ParseFactor() {
  if (Look().tag == Tag::LeftParenthesis) {
    Advance();
    ExpPtr x = ParseOr();
    Match(Tag::RightParenthesis);
    return x;
  } else if (Look().tag == Tag::Integer) {
    std::u32string v = Look().text;
    const Token& start = Look();
    Advance();
    return std::make_shared<ConstantExpression>(GetLoc(start), std::make_shared<Int32Type>(), v);
  } else if (Look().tag == Tag::Float) {
    std::u32string v = Look().text;
    const Token& start = Look();
    Advance();
    return std::make_shared<ConstantExpression>(GetLoc(start), std::make_shared<DoubleType>(), v);
  } else if (Look().tag == Tag::String) {
    std::u32string v = Look().text;
    const Token& start = Look();
    Advance();
    return std::make_shared<ConstantExpression>(GetLoc(start), std::make_shared<StringType>(), v);
  } else if (Look().tag == Tag::True) {
    const Token& start = Look();
    Advance();
    return std::make_shared<ConstantExpression>(GetLoc(start), std::make_shared<BooleanType>(), U"true");
  } else if (Look().tag == Tag::False) {
    const Token& start = Look();
    Advance();
    return std::make_shared<ConstantExpression>(GetLoc(start), std::make_shared<BooleanType>(), U"false");
  } else if (Look().tag == Tag::Identifier) {
    std::u32string name = Look().text;
    const Token& start = Look();
    Advance();
    return std::make_shared<ParameterExpression>(GetLoc(start), name, std::make_shared<UnknownType>());
  } else {
    throw ParserException(Look().line, Look().column, U"factor");
  }
}

ExpPtr Parser::ParseBlock() {
  const Token& start = Look();
  Match(Tag::LeftBrace);
  vector<ExpPtr> expressions;
  while (!IsEof() && Look().tag != Tag::RightBrace) {
    expressions.push_back(Statement());
  }
  Match(Tag::RightBrace);
  return std::make_shared<BlockExpression>(GetLoc(start), expressions);
}

ExpPtr Parser::IfStatement() {
  const Token& start = Look();
  Match(Tag::If);
  ExpPtr condition = ParseOr();
  ExpPtr ifTrue = ParseBlock();
  if (Look().tag == Tag::Else) {
    Match(Tag::Else);
    if (Look().tag == Tag::If) {
      auto chunk = IfStatement();
      return std::make_shared<ConditionalExpression>(GetLoc(start), condition, ifTrue, chunk);
    } else {
      auto chunk = ParseBlock();
      return std::make_shared<ConditionalExpression>(GetLoc(start), condition, ifTrue, chunk);
    }
  } else {
      auto empty = std::make_shared<DefaultExpression>(GetLoc(Look()));
      return std::make_shared<ConditionalExpression>(GetLoc(start), condition, ifTrue, empty);
  }
}

ExpPtr Parser::ParseVar() {
  const Token& start = Look();
  Match(Tag::Var);
  Token t = Match(Tag::Identifier);
  if (Look().tag == Tag::Colon) {
    Match(Tag::Colon);
    auto type = Optional<Ptr<TypeExpression>>(ParseType());
    if (Look().tag == Tag::Assign) {
      Match(Tag::Assign);
      auto value = Optional<Ptr<Ast>>(ParseOr());
      return New<Var>(GetPos(start), t.text, type, value);
    } else {
      return New<Var>(GetPos(start), t.text, type, Optional<Ptr<Ast>>());
    }
  } else {
    Match(Tag::Assign);
    auto value = Optional<Ptr<Ast>>(ParseOr());
    return New<Var>(GetPos(start), t.text, Optional<Ptr<TypeExpression>>(),
                    value);
  }
}

Ptr<Var> Parser::ParseVarDeclaration() {
  const Token& start = Look();
  Match(Tag::Var);
  Token t = Match(Tag::Identifier);
  Match(Tag::Colon);
  auto type = Optional<Ptr<TypeExpression>>(ParseType());
  return New<Var>(GetPos(start), t.text, type, Optional<Ptr<Ast>>());
}

Ptr<Def> Parser::ParseDef() {
  const Token& start = Look();
  Match(Tag::Def);
  String name = Match(Tag::Identifier).text;

  Match(Tag::LeftParenthesis);
  Vector<Parameter> parameters;
  if (Look().tag != Tag::RightParenthesis) {
    parameters.push_back(ParseParameter());
    while (!IsEof() && Look().tag != Tag::RightParenthesis) {
      Match(Tag::Comma);
      parameters.push_back(ParseParameter());
    }
  }
  Match(Tag::RightParenthesis);
  Match(Tag::Colon);
  auto returnType = ParseType();
  Ptr<Ast> body = ParseBlock();
  return New<Def>(GetPos(start), name, parameters,
                  Def::MakeFunctionType(GetPos(start), parameters, returnType),
                  body);
}

Parameter Parser::ParseParameter() {
  String name = Match(Tag::Identifier).text;
  Match(Tag::Colon);
  auto type = ParseType();
  return Parameter(name, type);
}

Ptr<TypeExpression> Parser::ParseType() {
  String name = Match(Tag::Identifier).text;
  if (Look().tag == Tag::LeftBracket) {
    Match(Tag::LeftBracket);
    const auto& token = Look();
    Vector<Ptr<TypeExpression>> types = ParseTypeArguments();
    auto result = New<TypeExpression>(GetPos(token), name, types);
    if (types.empty()) {
      throw ParserException(Look().line, Look().column, "type error");
    } else {
      return result;
    }
  } else {
    return New<TypeExpression>(GetPos(Look()), name);
  }
}

Ptr<Ast> Parser::ParseReturn() {
  const Token& start = Look();
  Match(Tag::Return);
  Ptr<Ast> value = ParseOr();
  return New<Return>(GetPos(start), value);
}

Vector<Ptr<TypeExpression>> Parser::ParseTypeArguments() {
  Match(Tag::LeftBracket);
  Vector<Ptr<TypeExpression>> types;
  types.push_back(ParseType());
  while (!IsEof() && Look().tag != Tag::RightBracket) {
    Match(Tag::Comma);
    types.push_back(ParseType());
  }
  Match(Tag::RightBracket);
  return types;
}

Ptr<While> Parser::ParseWhile() {
  const Token& start = Look();
  Token token = Match(Tag::While);
  Ptr<Ast> condition = ParseOr();
  Ptr<Ast> body = ParseBlock();
  return New<While>(GetPos(start), condition, body);
}

Ptr<DefClass> Parser::ParseDefClass() {
  const Token& start = Look();
  Match(Tag::Class);
  String name = Match(Tag::Identifier).text;
  Vector<Ptr<Var>> fields;
  Vector<Ptr<Def>> methods;
  Match(Tag::LeftBrace);
  while (!IsEof() && Look().tag != Tag::RightBrace) {
    Access access = ParseAccess();

    if (Look().tag == Tag::Var) {
      // ParseVar field: Type
      auto field = ParseVarDeclaration();
      field->access = access;
      fields.push_back(field);
    } else if (Look().tag == Tag::Def) {
      // def method(args..) { }
      auto method = ParseDef();
      method->access = access;
      methods.push_back(method);
    } else {
      throw ParserException(Look().line, Look().column, "unexpected token");
    }
  }
  Match(Tag::RightBrace);
  return New<DefClass>(GetPos(start), name, fields, methods);
}

Ptr<DefModule> Parser::ParseDefModule() {
  const Token& start = Look();
  Match(Tag::Module);
  String name = Match(Tag::Identifier).text;
  Vector<Ptr<Var>> fields;
  Vector<Ptr<Def>> methods;
  Match(Tag::LeftBrace);
  while (!IsEof() && Look().tag != Tag::RightBrace) {
    Access access = ParseAccess();
    if (Look().tag == Tag::Var) {
      // ParseVar field: Type
      auto field = ParseVarDeclaration();
      field->access = access;
      fields.push_back(field);
    } else if (Look().tag == Tag::Def) {
      // def method(args..) { }
      auto method = ParseDef();
      method->access = access;
      methods.push_back(method);
    } else {
      throw ParserException(Look().line, Look().column, "unexpected token");
    }
  }
  Match(Tag::RightBrace);
  return New<DefModule>(GetPos(start), name, fields, methods);
}

Access Parser::ParseAccess() {
  if (Look().tag == Tag::Private) {
    Match(Tag::Private);
    return Access::Private;
  } else if (Look().tag == Tag::Protected) {
    Match(Tag::Protected);
    return Access::Protected;
  } else {
    return Access::Public;
  }
}
} // namespace cygni
